{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "051af14f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "926a4471",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0               Adelpha_iphiclus\n",
       "1     Anartia_jatrophae_saturata\n",
       "2               Araschnia_levana\n",
       "3                 Auzakia_danava\n",
       "4                 Baeotus_beotus\n",
       "5          Catacroptera_cloanthe\n",
       "6                Chalinga_pratti\n",
       "7       Diaethria_gabaza_eupepla\n",
       "8            Doleschallia_melana\n",
       "9                  Eurema_blanda\n",
       "10           Hypolimnas_usambara\n",
       "11               Junonia_villida\n",
       "12             Kallima_paralekta\n",
       "13             Kallimoides_rumia\n",
       "14               Litinga_cottini\n",
       "15              Mallika_jacksoni\n",
       "16                Moduza_procris\n",
       "17               Parasarpa_zayla\n",
       "18            Phaedyma_columella\n",
       "19                Precis_pelarga\n",
       "20       Protogoniomorpha_temora\n",
       "21                 Salamis_cacta\n",
       "22             Smyrna_blomfildia\n",
       "23                Tacola_larymna\n",
       "24                   Yoma_algina\n",
       "25         Tacola_larymna_subset\n",
       "26            Yoma_algina_subset\n",
       "Name: ID, dtype: object"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get samples \n",
    "samples = pd.read_csv(\"../../config/samples.csv\")[\"ID\"]\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "de0a390f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Assembled sequence</th>\n",
       "      <th>N seqs</th>\n",
       "      <th>Sum length</th>\n",
       "      <th>Min. length</th>\n",
       "      <th>Avg. length</th>\n",
       "      <th>Max. length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adelpha_iphiclus</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,263</td>\n",
       "      <td>15,263</td>\n",
       "      <td>15,263</td>\n",
       "      <td>15,263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,281</td>\n",
       "      <td>15,281</td>\n",
       "      <td>15,281</td>\n",
       "      <td>15,281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Araschnia_levana</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,207</td>\n",
       "      <td>15,207</td>\n",
       "      <td>15,207</td>\n",
       "      <td>15,207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>15,017</td>\n",
       "      <td>3,204</td>\n",
       "      <td>5,005.7</td>\n",
       "      <td>6,944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Baeotus_beotus</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,131</td>\n",
       "      <td>15,131</td>\n",
       "      <td>15,131</td>\n",
       "      <td>15,131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13,058</td>\n",
       "      <td>13,058</td>\n",
       "      <td>13,058</td>\n",
       "      <td>13,058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Chalinga_pratti</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,244</td>\n",
       "      <td>15,244</td>\n",
       "      <td>15,244</td>\n",
       "      <td>15,244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Diaethria_gabaza_eupepla</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,156</td>\n",
       "      <td>15,156</td>\n",
       "      <td>15,156</td>\n",
       "      <td>15,156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Doleschallia_melana</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,379</td>\n",
       "      <td>15,379</td>\n",
       "      <td>15,379</td>\n",
       "      <td>15,379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Eurema_blanda</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,123</td>\n",
       "      <td>15,123</td>\n",
       "      <td>15,123</td>\n",
       "      <td>15,123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Hypolimnas_usambara</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,158</td>\n",
       "      <td>15,158</td>\n",
       "      <td>15,158</td>\n",
       "      <td>15,158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>13,451</td>\n",
       "      <td>1,712</td>\n",
       "      <td>3,362.8</td>\n",
       "      <td>6,323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Kallima_paralekta</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,200</td>\n",
       "      <td>15,200</td>\n",
       "      <td>15,200</td>\n",
       "      <td>15,200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Kallimoides_rumia</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,234</td>\n",
       "      <td>15,234</td>\n",
       "      <td>15,234</td>\n",
       "      <td>15,234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Litinga_cottini</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,189</td>\n",
       "      <td>15,189</td>\n",
       "      <td>15,189</td>\n",
       "      <td>15,189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Mallika_jacksoni</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,193</td>\n",
       "      <td>15,193</td>\n",
       "      <td>15,193</td>\n",
       "      <td>15,193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Moduza_procris</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,195</td>\n",
       "      <td>15,195</td>\n",
       "      <td>15,195</td>\n",
       "      <td>15,195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Parasarpa_zayla</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,275</td>\n",
       "      <td>15,275</td>\n",
       "      <td>15,275</td>\n",
       "      <td>15,275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Phaedyma_columella</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,197</td>\n",
       "      <td>15,197</td>\n",
       "      <td>15,197</td>\n",
       "      <td>15,197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Precis_pelarga</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,224</td>\n",
       "      <td>15,224</td>\n",
       "      <td>15,224</td>\n",
       "      <td>15,224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Protogoniomorpha_temora</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,335</td>\n",
       "      <td>15,335</td>\n",
       "      <td>15,335</td>\n",
       "      <td>15,335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Salamis_cacta</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,218</td>\n",
       "      <td>15,218</td>\n",
       "      <td>15,218</td>\n",
       "      <td>15,218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Smyrna_blomfildia</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,149</td>\n",
       "      <td>15,149</td>\n",
       "      <td>15,149</td>\n",
       "      <td>15,149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Tacola_larymna</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,475</td>\n",
       "      <td>15,475</td>\n",
       "      <td>15,475</td>\n",
       "      <td>15,475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Yoma_algina</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15,231</td>\n",
       "      <td>15,231</td>\n",
       "      <td>15,231</td>\n",
       "      <td>15,231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Tacola_larymna_subset</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Yoma_algina_subset</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            ID Assembled sequence  N seqs Sum length  \\\n",
       "0             Adelpha_iphiclus               True     1.0     15,263   \n",
       "1   Anartia_jatrophae_saturata               True     1.0     15,281   \n",
       "2             Araschnia_levana               True     1.0     15,207   \n",
       "3               Auzakia_danava               True     3.0     15,017   \n",
       "4               Baeotus_beotus               True     1.0     15,131   \n",
       "5        Catacroptera_cloanthe               True     1.0     13,058   \n",
       "6              Chalinga_pratti               True     1.0     15,244   \n",
       "7     Diaethria_gabaza_eupepla               True     1.0     15,156   \n",
       "8          Doleschallia_melana               True     1.0     15,379   \n",
       "9                Eurema_blanda               True     1.0     15,123   \n",
       "10         Hypolimnas_usambara               True     1.0     15,158   \n",
       "11             Junonia_villida               True     4.0     13,451   \n",
       "12           Kallima_paralekta               True     1.0     15,200   \n",
       "13           Kallimoides_rumia               True     1.0     15,234   \n",
       "14             Litinga_cottini               True     1.0     15,189   \n",
       "15            Mallika_jacksoni               True     1.0     15,193   \n",
       "16              Moduza_procris               True     1.0     15,195   \n",
       "17             Parasarpa_zayla               True     1.0     15,275   \n",
       "18          Phaedyma_columella               True     1.0     15,197   \n",
       "19              Precis_pelarga               True     1.0     15,224   \n",
       "20     Protogoniomorpha_temora               True     1.0     15,335   \n",
       "21               Salamis_cacta               True     1.0     15,218   \n",
       "22           Smyrna_blomfildia               True     1.0     15,149   \n",
       "23              Tacola_larymna               True     1.0     15,475   \n",
       "24                 Yoma_algina               True     1.0     15,231   \n",
       "25       Tacola_larymna_subset              False     NaN        NaN   \n",
       "26          Yoma_algina_subset              False     NaN        NaN   \n",
       "\n",
       "   Min. length Avg. length Max. length  \n",
       "0       15,263      15,263      15,263  \n",
       "1       15,281      15,281      15,281  \n",
       "2       15,207      15,207      15,207  \n",
       "3        3,204     5,005.7       6,944  \n",
       "4       15,131      15,131      15,131  \n",
       "5       13,058      13,058      13,058  \n",
       "6       15,244      15,244      15,244  \n",
       "7       15,156      15,156      15,156  \n",
       "8       15,379      15,379      15,379  \n",
       "9       15,123      15,123      15,123  \n",
       "10      15,158      15,158      15,158  \n",
       "11       1,712     3,362.8       6,323  \n",
       "12      15,200      15,200      15,200  \n",
       "13      15,234      15,234      15,234  \n",
       "14      15,189      15,189      15,189  \n",
       "15      15,193      15,193      15,193  \n",
       "16      15,195      15,195      15,195  \n",
       "17      15,275      15,275      15,275  \n",
       "18      15,197      15,197      15,197  \n",
       "19      15,224      15,224      15,224  \n",
       "20      15,335      15,335      15,335  \n",
       "21      15,218      15,218      15,218  \n",
       "22      15,149      15,149      15,149  \n",
       "23      15,475      15,475      15,475  \n",
       "24      15,231      15,231      15,231  \n",
       "25         NaN         NaN         NaN  \n",
       "26         NaN         NaN         NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqkit_dir = \"../../results/seqkit/\"\n",
    "\n",
    "# list to hold dataframes\n",
    "list_seqkit = []\n",
    "# iterate throught seqkit results files\n",
    "for file in os.listdir(seqkit_dir):\n",
    "    if file.endswith(\".txt\"):\n",
    "        # use regex separator for variable whitespace with python engine\n",
    "        df = pd.read_csv(os.path.join(seqkit_dir, file), sep=r\"\\s+\")\n",
    "        # append to list\n",
    "        list_seqkit.append(df)\n",
    "# concat list\n",
    "dat_seqkit = pd.concat(list_seqkit, ignore_index=True)\n",
    "# clean up dataframe\n",
    "dat_seqkit.drop(\"format\", axis=1, inplace=True)\n",
    "dat_seqkit.drop(\"type\", axis=1, inplace=True)\n",
    "dat_seqkit[\"ID\"] = dat_seqkit[\"file\"].str.replace(\".fasta\", \"\", regex=False)\n",
    "dat_seqkit.drop(\"file\", axis=1, inplace=True)\n",
    "dat_seqkit.rename(columns={\"num_seqs\": \"N seqs\", \n",
    "                           \"sum_len\": \"Sum length\", \n",
    "                           \"min_len\": \"Min. length\", \n",
    "                           \"avg_len\": \"Avg. length\", \n",
    "                           \"max_len\": \"Max. length\"}, inplace=True)\n",
    "# join with sample names\n",
    "dat_seqkit = pd.merge(samples, dat_seqkit, how=\"left\", on=\"ID\")\n",
    "# add column to state if assembly present\n",
    "dat_seqkit[\"Assembled sequence\"] = dat_seqkit[\"N seqs\"].apply(lambda x: \"True\" if x > 0 else \"False\")\n",
    "dat_seqkit = dat_seqkit[[\"ID\", \"Assembled sequence\", \"N seqs\", \"Sum length\", \"Min. length\", \"Avg. length\", \"Max. length\"]]\n",
    "dat_seqkit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "116e6ec5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Assembled sequence</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Length</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>GC content</th>\n",
       "      <th>Superkingdom</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adelpha_iphiclus</td>\n",
       "      <td>True</td>\n",
       "      <td>Adelpha_iphiclus_circular</td>\n",
       "      <td>15263.0</td>\n",
       "      <td>17.0734</td>\n",
       "      <td>0.1946</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Adelpha iphiclus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>True</td>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>15281.0</td>\n",
       "      <td>17.0647</td>\n",
       "      <td>0.1862</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Yoma algina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Araschnia_levana</td>\n",
       "      <td>True</td>\n",
       "      <td>Araschnia_levana_circular</td>\n",
       "      <td>15207.0</td>\n",
       "      <td>17.0016</td>\n",
       "      <td>0.1835</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Araschnia levana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>True</td>\n",
       "      <td>Auzakia_danava_contig0</td>\n",
       "      <td>6944.0</td>\n",
       "      <td>16.5379</td>\n",
       "      <td>0.2036</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Auzakia danava</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>True</td>\n",
       "      <td>Auzakia_danava_contig1</td>\n",
       "      <td>4869.0</td>\n",
       "      <td>17.5841</td>\n",
       "      <td>0.1818</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Auzakia danava</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>True</td>\n",
       "      <td>Auzakia_danava_contig2</td>\n",
       "      <td>3204.0</td>\n",
       "      <td>18.4320</td>\n",
       "      <td>0.1898</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Auzakia danava</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Baeotus_beotus</td>\n",
       "      <td>True</td>\n",
       "      <td>Baeotus_beotus_circular</td>\n",
       "      <td>15131.0</td>\n",
       "      <td>17.0449</td>\n",
       "      <td>0.1954</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Baeotus beotus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>True</td>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>13058.0</td>\n",
       "      <td>17.4831</td>\n",
       "      <td>0.2261</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Catacroptera cloanthe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Chalinga_pratti</td>\n",
       "      <td>True</td>\n",
       "      <td>Chalinga_pratti_contig0</td>\n",
       "      <td>15244.0</td>\n",
       "      <td>17.1300</td>\n",
       "      <td>0.1848</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Chalinga pratti</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Diaethria_gabaza_eupepla</td>\n",
       "      <td>True</td>\n",
       "      <td>Diaethria_gabaza_eupepla_circular</td>\n",
       "      <td>15156.0</td>\n",
       "      <td>17.0368</td>\n",
       "      <td>0.1951</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Ariadne ariadne</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Doleschallia_melana</td>\n",
       "      <td>True</td>\n",
       "      <td>Doleschallia_melana_contig0</td>\n",
       "      <td>15379.0</td>\n",
       "      <td>16.9878</td>\n",
       "      <td>0.1997</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Doleschallia melana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Eurema_blanda</td>\n",
       "      <td>True</td>\n",
       "      <td>Eurema_blanda_circular</td>\n",
       "      <td>15123.0</td>\n",
       "      <td>17.0998</td>\n",
       "      <td>0.1943</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Pieridae</td>\n",
       "      <td>Eurema blanda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Hypolimnas_usambara</td>\n",
       "      <td>True</td>\n",
       "      <td>Hypolimnas_usambara_contig0</td>\n",
       "      <td>15158.0</td>\n",
       "      <td>17.0581</td>\n",
       "      <td>0.2105</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Hypolimnas usambara</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>True</td>\n",
       "      <td>Junonia_villida_contig0</td>\n",
       "      <td>2160.0</td>\n",
       "      <td>17.3014</td>\n",
       "      <td>0.1727</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Junonia vestina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>True</td>\n",
       "      <td>Junonia_villida_contig1</td>\n",
       "      <td>1712.0</td>\n",
       "      <td>17.2442</td>\n",
       "      <td>0.1828</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Junonia villida</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>True</td>\n",
       "      <td>Junonia_villida_contig2</td>\n",
       "      <td>6323.0</td>\n",
       "      <td>16.8507</td>\n",
       "      <td>0.2121</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Junonia villida</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>True</td>\n",
       "      <td>Junonia_villida_contig3</td>\n",
       "      <td>3256.0</td>\n",
       "      <td>19.0310</td>\n",
       "      <td>0.2006</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Junonia villida</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Kallima_paralekta</td>\n",
       "      <td>True</td>\n",
       "      <td>Kallima_paralekta_circular</td>\n",
       "      <td>15200.0</td>\n",
       "      <td>17.0488</td>\n",
       "      <td>0.2047</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Kallima paralekta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Kallimoides_rumia</td>\n",
       "      <td>True</td>\n",
       "      <td>Kallimoides_rumia_circular</td>\n",
       "      <td>15234.0</td>\n",
       "      <td>17.1451</td>\n",
       "      <td>0.1910</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Kallimoides rumia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Litinga_cottini</td>\n",
       "      <td>True</td>\n",
       "      <td>Litinga_cottini_contig0</td>\n",
       "      <td>15189.0</td>\n",
       "      <td>16.9812</td>\n",
       "      <td>0.1970</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Limenitis cottini</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Mallika_jacksoni</td>\n",
       "      <td>True</td>\n",
       "      <td>Mallika_jacksoni_circular</td>\n",
       "      <td>15193.0</td>\n",
       "      <td>16.9918</td>\n",
       "      <td>0.2061</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Mallika jacksoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Moduza_procris</td>\n",
       "      <td>True</td>\n",
       "      <td>Moduza_procris_circular</td>\n",
       "      <td>15195.0</td>\n",
       "      <td>17.0083</td>\n",
       "      <td>0.1932</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Moduza procris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Parasarpa_zayla</td>\n",
       "      <td>True</td>\n",
       "      <td>Parasarpa_zayla_circular</td>\n",
       "      <td>15275.0</td>\n",
       "      <td>17.0816</td>\n",
       "      <td>0.1968</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Parasarpa zayla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Phaedyma_columella</td>\n",
       "      <td>True</td>\n",
       "      <td>Phaedyma_columella_circular</td>\n",
       "      <td>15197.0</td>\n",
       "      <td>17.0053</td>\n",
       "      <td>0.1969</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Phaedyma columella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Precis_pelarga</td>\n",
       "      <td>True</td>\n",
       "      <td>Precis_pelarga_circular</td>\n",
       "      <td>15224.0</td>\n",
       "      <td>16.9779</td>\n",
       "      <td>0.1992</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Precis pelarga</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Protogoniomorpha_temora</td>\n",
       "      <td>True</td>\n",
       "      <td>Protogoniomorpha_temora_contig0</td>\n",
       "      <td>15335.0</td>\n",
       "      <td>17.0245</td>\n",
       "      <td>0.1933</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Protogoniomorpha temora</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Salamis_cacta</td>\n",
       "      <td>True</td>\n",
       "      <td>Salamis_cacta_contig0</td>\n",
       "      <td>15218.0</td>\n",
       "      <td>16.9870</td>\n",
       "      <td>0.1927</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Salamis cacta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Smyrna_blomfildia</td>\n",
       "      <td>True</td>\n",
       "      <td>Smyrna_blomfildia_circular</td>\n",
       "      <td>15149.0</td>\n",
       "      <td>17.0302</td>\n",
       "      <td>0.2050</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Nymphalis urticae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Tacola_larymna</td>\n",
       "      <td>True</td>\n",
       "      <td>Tacola_larymna_contig0</td>\n",
       "      <td>15475.0</td>\n",
       "      <td>17.0210</td>\n",
       "      <td>0.2005</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Tacola larymna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Yoma_algina</td>\n",
       "      <td>True</td>\n",
       "      <td>Yoma_algina_circular</td>\n",
       "      <td>15231.0</td>\n",
       "      <td>17.1401</td>\n",
       "      <td>0.1870</td>\n",
       "      <td>no-hit</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Lepidoptera</td>\n",
       "      <td>Nymphalidae</td>\n",
       "      <td>Yoma algina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Tacola_larymna_subset</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Yoma_algina_subset</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            ID Assembled sequence  \\\n",
       "0             Adelpha_iphiclus               True   \n",
       "1   Anartia_jatrophae_saturata               True   \n",
       "2             Araschnia_levana               True   \n",
       "3               Auzakia_danava               True   \n",
       "4               Auzakia_danava               True   \n",
       "5               Auzakia_danava               True   \n",
       "6               Baeotus_beotus               True   \n",
       "7        Catacroptera_cloanthe               True   \n",
       "8              Chalinga_pratti               True   \n",
       "9     Diaethria_gabaza_eupepla               True   \n",
       "10         Doleschallia_melana               True   \n",
       "11               Eurema_blanda               True   \n",
       "12         Hypolimnas_usambara               True   \n",
       "13             Junonia_villida               True   \n",
       "14             Junonia_villida               True   \n",
       "15             Junonia_villida               True   \n",
       "16             Junonia_villida               True   \n",
       "17           Kallima_paralekta               True   \n",
       "18           Kallimoides_rumia               True   \n",
       "19             Litinga_cottini               True   \n",
       "20            Mallika_jacksoni               True   \n",
       "21              Moduza_procris               True   \n",
       "22             Parasarpa_zayla               True   \n",
       "23          Phaedyma_columella               True   \n",
       "24              Precis_pelarga               True   \n",
       "25     Protogoniomorpha_temora               True   \n",
       "26               Salamis_cacta               True   \n",
       "27           Smyrna_blomfildia               True   \n",
       "28              Tacola_larymna               True   \n",
       "29                 Yoma_algina               True   \n",
       "30       Tacola_larymna_subset              False   \n",
       "31          Yoma_algina_subset              False   \n",
       "\n",
       "                                Contig   Length  Coverage  GC content  \\\n",
       "0            Adelpha_iphiclus_circular  15263.0   17.0734      0.1946   \n",
       "1   Anartia_jatrophae_saturata_contig0  15281.0   17.0647      0.1862   \n",
       "2            Araschnia_levana_circular  15207.0   17.0016      0.1835   \n",
       "3               Auzakia_danava_contig0   6944.0   16.5379      0.2036   \n",
       "4               Auzakia_danava_contig1   4869.0   17.5841      0.1818   \n",
       "5               Auzakia_danava_contig2   3204.0   18.4320      0.1898   \n",
       "6              Baeotus_beotus_circular  15131.0   17.0449      0.1954   \n",
       "7        Catacroptera_cloanthe_contig0  13058.0   17.4831      0.2261   \n",
       "8              Chalinga_pratti_contig0  15244.0   17.1300      0.1848   \n",
       "9    Diaethria_gabaza_eupepla_circular  15156.0   17.0368      0.1951   \n",
       "10         Doleschallia_melana_contig0  15379.0   16.9878      0.1997   \n",
       "11              Eurema_blanda_circular  15123.0   17.0998      0.1943   \n",
       "12         Hypolimnas_usambara_contig0  15158.0   17.0581      0.2105   \n",
       "13             Junonia_villida_contig0   2160.0   17.3014      0.1727   \n",
       "14             Junonia_villida_contig1   1712.0   17.2442      0.1828   \n",
       "15             Junonia_villida_contig2   6323.0   16.8507      0.2121   \n",
       "16             Junonia_villida_contig3   3256.0   19.0310      0.2006   \n",
       "17          Kallima_paralekta_circular  15200.0   17.0488      0.2047   \n",
       "18          Kallimoides_rumia_circular  15234.0   17.1451      0.1910   \n",
       "19             Litinga_cottini_contig0  15189.0   16.9812      0.1970   \n",
       "20           Mallika_jacksoni_circular  15193.0   16.9918      0.2061   \n",
       "21             Moduza_procris_circular  15195.0   17.0083      0.1932   \n",
       "22            Parasarpa_zayla_circular  15275.0   17.0816      0.1968   \n",
       "23         Phaedyma_columella_circular  15197.0   17.0053      0.1969   \n",
       "24             Precis_pelarga_circular  15224.0   16.9779      0.1992   \n",
       "25     Protogoniomorpha_temora_contig0  15335.0   17.0245      0.1933   \n",
       "26               Salamis_cacta_contig0  15218.0   16.9870      0.1927   \n",
       "27          Smyrna_blomfildia_circular  15149.0   17.0302      0.2050   \n",
       "28              Tacola_larymna_contig0  15475.0   17.0210      0.2005   \n",
       "29                Yoma_algina_circular  15231.0   17.1401      0.1870   \n",
       "30                                 NaN      NaN       NaN         NaN   \n",
       "31                                 NaN      NaN       NaN         NaN   \n",
       "\n",
       "   Superkingdom  Kingdom      Phylum    Class        Order       Family  \\\n",
       "0        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "1        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "2        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "3        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "4        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "5        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "6        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "7        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "8        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "9        no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "10       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "11       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera     Pieridae   \n",
       "12       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "13       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "14       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "15       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "16       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "17       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "18       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "19       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "20       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "21       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "22       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "23       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "24       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "25       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "26       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "27       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "28       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "29       no-hit  Metazoa  Arthropoda  Insecta  Lepidoptera  Nymphalidae   \n",
       "30          NaN      NaN         NaN      NaN          NaN          NaN   \n",
       "31          NaN      NaN         NaN      NaN          NaN          NaN   \n",
       "\n",
       "                    Species  \n",
       "0          Adelpha iphiclus  \n",
       "1               Yoma algina  \n",
       "2          Araschnia levana  \n",
       "3            Auzakia danava  \n",
       "4            Auzakia danava  \n",
       "5            Auzakia danava  \n",
       "6            Baeotus beotus  \n",
       "7     Catacroptera cloanthe  \n",
       "8           Chalinga pratti  \n",
       "9           Ariadne ariadne  \n",
       "10      Doleschallia melana  \n",
       "11            Eurema blanda  \n",
       "12      Hypolimnas usambara  \n",
       "13          Junonia vestina  \n",
       "14          Junonia villida  \n",
       "15          Junonia villida  \n",
       "16          Junonia villida  \n",
       "17        Kallima paralekta  \n",
       "18        Kallimoides rumia  \n",
       "19        Limenitis cottini  \n",
       "20         Mallika jacksoni  \n",
       "21           Moduza procris  \n",
       "22          Parasarpa zayla  \n",
       "23       Phaedyma columella  \n",
       "24           Precis pelarga  \n",
       "25  Protogoniomorpha temora  \n",
       "26            Salamis cacta  \n",
       "27        Nymphalis urticae  \n",
       "28           Tacola larymna  \n",
       "29              Yoma algina  \n",
       "30                      NaN  \n",
       "31                      NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blobtools_dir = \"../../results/blobtools/\"\n",
    "\n",
    "# list to hold dataframes\n",
    "list_blobtools = []\n",
    "# iterate throught blobtools results files\n",
    "for directory in os.listdir(blobtools_dir):\n",
    "    if os.path.isdir(os.path.join(blobtools_dir, directory)):\n",
    "        for file in os.listdir(os.path.join(blobtools_dir, directory)):\n",
    "            if file == \"table.tsv\":\n",
    "                file_bt = pd.read_csv(os.path.join(blobtools_dir, directory, file), sep=\"\\t\")\n",
    "                # find column ending with \"_cov\" and rename to \"coverage\"\n",
    "                cov_cols = [col for col in file_bt.columns if col.endswith(\"_cov\")]\n",
    "                if cov_cols:\n",
    "                    file_bt.rename(columns={cov_cols[0]: \"coverage\"}, inplace=True)\n",
    "                # add sample ID from directory name\n",
    "                file_bt['ID'] = directory\n",
    "                file_bt.drop(\"index\", axis=1, inplace=True)\n",
    "                list_blobtools.append(file_bt)\n",
    "# concat list\n",
    "dat_blobtools = pd.concat(list_blobtools, ignore_index=True)\n",
    "# clean up dataframe\n",
    "dat_blobtools.rename(columns={\n",
    "    \"identifiers\" : \"Contig\", \n",
    "    \"length\" : \"Length\",\n",
    "    \"coverage\" : \"Coverage\",\n",
    "    \"gc\" : \"GC content\",\n",
    "    \"bestsumorder_superkingdom\" : \"Superkingdom\",\n",
    "    \"bestsumorder_kingdom\" : \"Kingdom\",\n",
    "    \"bestsumorder_phylum\" : \"Phylum\",\n",
    "    \"bestsumorder_class\" : \"Class\",\n",
    "    \"bestsumorder_order\" : \"Order\",\n",
    "    \"bestsumorder_family\" : \"Family\",\n",
    "    \"bestsumorder_species\" : \"Species\"\n",
    "}, inplace=True)\n",
    "\n",
    "# join with sample names\n",
    "dat_blobtools = pd.merge(samples, dat_blobtools, how=\"left\", on=\"ID\")\n",
    "# add column to state if assembly present\n",
    "dat_blobtools[\"Assembled sequence\"] = dat_blobtools[\"Contig\"].apply(lambda x: \"False\" if pd.isna(x) else \"True\")\n",
    "\n",
    "dat_blobtools = dat_blobtools[[\"ID\", \"Assembled sequence\", \"Contig\", \"Length\", \"Coverage\", \"GC content\", \"Superkingdom\", \"Kingdom\", \"Phylum\", \"Class\", \"Order\", \"Family\", \"Species\"]]\n",
    "dat_blobtools\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "344c33b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>4</td>\n",
       "      <td>1502</td>\n",
       "      <td>nad5_0</td>\n",
       "      <td>237982419.4</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>1586</td>\n",
       "      <td>2922</td>\n",
       "      <td>nad4</td>\n",
       "      <td>214655501.8</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>2921</td>\n",
       "      <td>3206</td>\n",
       "      <td>nad4l</td>\n",
       "      <td>4178913.3</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>3369</td>\n",
       "      <td>3867</td>\n",
       "      <td>nad6</td>\n",
       "      <td>3279670.2</td>\n",
       "      <td>+</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>3895</td>\n",
       "      <td>5047</td>\n",
       "      <td>cob</td>\n",
       "      <td>323602426.3</td>\n",
       "      <td>+</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>8669</td>\n",
       "      <td>9347</td>\n",
       "      <td>atp6</td>\n",
       "      <td>35353613.3</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>922</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>9340</td>\n",
       "      <td>9502</td>\n",
       "      <td>atp8</td>\n",
       "      <td>65220.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>925</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>9603</td>\n",
       "      <td>10317</td>\n",
       "      <td>cox2</td>\n",
       "      <td>107385243.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>10379</td>\n",
       "      <td>11915</td>\n",
       "      <td>cox1</td>\n",
       "      <td>566136198.5</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>931</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>12111</td>\n",
       "      <td>13056</td>\n",
       "      <td>nad2</td>\n",
       "      <td>61252546.4</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>397 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Contig  Start    End    Name        Score  \\\n",
       "0    Anartia_jatrophae_saturata_contig0      4   1502  nad5_0  237982419.4   \n",
       "2    Anartia_jatrophae_saturata_contig0   1586   2922    nad4  214655501.8   \n",
       "3    Anartia_jatrophae_saturata_contig0   2921   3206   nad4l    4178913.3   \n",
       "6    Anartia_jatrophae_saturata_contig0   3369   3867    nad6    3279670.2   \n",
       "7    Anartia_jatrophae_saturata_contig0   3895   5047     cob  323602426.3   \n",
       "..                                  ...    ...    ...     ...          ...   \n",
       "921       Catacroptera_cloanthe_contig0   8669   9347    atp6   35353613.3   \n",
       "922       Catacroptera_cloanthe_contig0   9340   9502    atp8      65220.0   \n",
       "925       Catacroptera_cloanthe_contig0   9603  10317    cox2  107385243.0   \n",
       "927       Catacroptera_cloanthe_contig0  10379  11915    cox1  566136198.5   \n",
       "931       Catacroptera_cloanthe_contig0  12111  13056    nad2   61252546.4   \n",
       "\n",
       "    Strand                          ID  \n",
       "0        -  Anartia_jatrophae_saturata  \n",
       "2        -  Anartia_jatrophae_saturata  \n",
       "3        -  Anartia_jatrophae_saturata  \n",
       "6        +  Anartia_jatrophae_saturata  \n",
       "7        +  Anartia_jatrophae_saturata  \n",
       "..     ...                         ...  \n",
       "921      -       Catacroptera_cloanthe  \n",
       "922      -       Catacroptera_cloanthe  \n",
       "925      -       Catacroptera_cloanthe  \n",
       "927      -       Catacroptera_cloanthe  \n",
       "931      -       Catacroptera_cloanthe  \n",
       "\n",
       "[397 rows x 7 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations_dir = \"../../results/annotations/\"\n",
    "\n",
    "list_bed = []\n",
    "\n",
    "# iterate throught annotations results files\n",
    "for file in os.listdir(annotations_dir): \n",
    "    if os.path.isdir(os.path.join(annotations_dir, file)):\n",
    "        for (root,dirs,files) in os.walk(os.path.join(annotations_dir, file), topdown=True):\n",
    "            for f in files: \n",
    "                if f == \"result.bed\":\n",
    "                    dat_ann = pd.read_csv(os.path.join(root, f), sep=\"\\t\", header=None, names=[\"Contig\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"])\n",
    "                    # add sample name to dataframe\n",
    "                    dat_ann[\"ID\"] = file\n",
    "                    # append to list\n",
    "                    list_bed.append(dat_ann)\n",
    "# concat list\n",
    "dat_bed = pd.concat(list_bed, ignore_index=True)\n",
    "\n",
    "# remove annotations for tRNAs and origin of replication\n",
    "dat_bed = dat_bed[~dat_bed[\"Name\"].str.startswith((\"trn\", \"OH\", \"OL\"))]\n",
    "\n",
    "# join with sample names\n",
    "# dat_bed = pd.merge(samples, dat_bed, how=\"left\", on=\"ID\")\n",
    "\n",
    "# add column to state if assembly present\n",
    "#dat_bed[\"Assembled sequence\"] = dat_bed[\"Contig\"].apply(lambda x: \"False\" if pd.isna(x) else \"True\")\n",
    "\n",
    "dat_bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "4bfc42c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>N annotations</th>\n",
       "      <th>N cox1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adelpha_iphiclus</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Araschnia_levana</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Baeotus_beotus</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Chalinga_pratti</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Diaethria_gabaza_eupepla</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Doleschallia_melana</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Eurema_blanda</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Hypolimnas_usambara</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Kallima_paralekta</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Kallimoides_rumia</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Litinga_cottini</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Mallika_jacksoni</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Moduza_procris</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Parasarpa_zayla</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Phaedyma_columella</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Precis_pelarga</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Protogoniomorpha_temora</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Salamis_cacta</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Smyrna_blomfildia</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Tacola_larymna</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Yoma_algina</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            ID  N annotations  N cox1\n",
       "0             Adelpha_iphiclus             16       1\n",
       "1   Anartia_jatrophae_saturata             17       1\n",
       "2             Araschnia_levana             15       1\n",
       "3               Auzakia_danava             18       1\n",
       "4               Baeotus_beotus             15       1\n",
       "5        Catacroptera_cloanthe             14       1\n",
       "6              Chalinga_pratti             15       1\n",
       "7     Diaethria_gabaza_eupepla             15       1\n",
       "8          Doleschallia_melana             17       1\n",
       "9                Eurema_blanda             15       1\n",
       "10         Hypolimnas_usambara             18       1\n",
       "11             Junonia_villida             15       1\n",
       "12           Kallima_paralekta             16       1\n",
       "13           Kallimoides_rumia             15       1\n",
       "14             Litinga_cottini             17       1\n",
       "15            Mallika_jacksoni             16       1\n",
       "16              Moduza_procris             16       1\n",
       "17             Parasarpa_zayla             16       1\n",
       "18          Phaedyma_columella             16       1\n",
       "19              Precis_pelarga             17       1\n",
       "20     Protogoniomorpha_temora             16       1\n",
       "21               Salamis_cacta             17       1\n",
       "22           Smyrna_blomfildia             15       1\n",
       "23              Tacola_larymna             15       1\n",
       "24                 Yoma_algina             15       1"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get gene counts per sample\n",
    "gene_counts = dat_bed.groupby([\"ID\"])[[\"Name\"]].count().reset_index()\n",
    "\n",
    "# get the numer of \"cox1\" annotations per sample\n",
    "cox1_counts = dat_bed.groupby([\"ID\"])[\"Name\"].apply(lambda x: (x == \"cox1\").sum())\n",
    "\n",
    "# combine gene counts and cox1 counts into a single dataframe\n",
    "dat_annotations = pd.merge(gene_counts, cox1_counts, how=\"outer\", on=\"ID\").rename(columns={\"Name_x\": \"N annotations\", \"Name_y\": \"N cox1\"})\n",
    "dat_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "82c50d93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name</th>\n",
       "      <th>Present_mafft_filtered</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>3896</td>\n",
       "      <td>5047</td>\n",
       "      <td>+</td>\n",
       "      <td>cob</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chalinga_pratti</td>\n",
       "      <td>Chalinga_pratti_contig0</td>\n",
       "      <td>9205</td>\n",
       "      <td>10359</td>\n",
       "      <td>-</td>\n",
       "      <td>cob</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Protogoniomorpha_temora</td>\n",
       "      <td>Protogoniomorpha_temora_contig0</td>\n",
       "      <td>8938</td>\n",
       "      <td>10089</td>\n",
       "      <td>-</td>\n",
       "      <td>cob</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Parasarpa_zayla</td>\n",
       "      <td>Parasarpa_zayla_circular</td>\n",
       "      <td>9469</td>\n",
       "      <td>10617</td>\n",
       "      <td>-</td>\n",
       "      <td>cob</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>Junonia_villida_contig3</td>\n",
       "      <td>269</td>\n",
       "      <td>1420</td>\n",
       "      <td>+</td>\n",
       "      <td>cob</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>Precis_pelarga</td>\n",
       "      <td>Precis_pelarga_circular</td>\n",
       "      <td>4749</td>\n",
       "      <td>5426</td>\n",
       "      <td>+</td>\n",
       "      <td>atp6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>Salamis_cacta</td>\n",
       "      <td>Salamis_cacta_contig0</td>\n",
       "      <td>2970</td>\n",
       "      <td>3647</td>\n",
       "      <td>-</td>\n",
       "      <td>atp6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>Doleschallia_melana</td>\n",
       "      <td>Doleschallia_melana_contig0</td>\n",
       "      <td>9348</td>\n",
       "      <td>10025</td>\n",
       "      <td>+</td>\n",
       "      <td>atp6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>Litinga_cottini</td>\n",
       "      <td>Litinga_cottini_contig0</td>\n",
       "      <td>9139</td>\n",
       "      <td>9816</td>\n",
       "      <td>+</td>\n",
       "      <td>atp6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>8670</td>\n",
       "      <td>9347</td>\n",
       "      <td>-</td>\n",
       "      <td>atp6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>370 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             ID                              Contig Start  \\\n",
       "0    Anartia_jatrophae_saturata  Anartia_jatrophae_saturata_contig0  3896   \n",
       "1               Chalinga_pratti             Chalinga_pratti_contig0  9205   \n",
       "2       Protogoniomorpha_temora     Protogoniomorpha_temora_contig0  8938   \n",
       "3               Parasarpa_zayla            Parasarpa_zayla_circular  9469   \n",
       "4               Junonia_villida             Junonia_villida_contig3   269   \n",
       "..                          ...                                 ...   ...   \n",
       "365              Precis_pelarga             Precis_pelarga_circular  4749   \n",
       "366               Salamis_cacta               Salamis_cacta_contig0  2970   \n",
       "367         Doleschallia_melana         Doleschallia_melana_contig0  9348   \n",
       "368             Litinga_cottini             Litinga_cottini_contig0  9139   \n",
       "369       Catacroptera_cloanthe       Catacroptera_cloanthe_contig0  8670   \n",
       "\n",
       "       End Strand  Name Present_mafft_filtered  \n",
       "0     5047      +   cob                   True  \n",
       "1    10359      -   cob                   True  \n",
       "2    10089      -   cob                   True  \n",
       "3    10617      -   cob                   True  \n",
       "4     1420      +   cob                   True  \n",
       "..     ...    ...   ...                    ...  \n",
       "365   5426      +  atp6                   True  \n",
       "366   3647      -  atp6                   True  \n",
       "367  10025      +  atp6                   True  \n",
       "368   9816      +  atp6                   True  \n",
       "369   9347      -  atp6                   True  \n",
       "\n",
       "[370 rows x 7 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# input directory for annotated sequences\n",
    "extract_alignments_dir = \"../../results/mafft_filtered/\"\n",
    "\n",
    "# functions\n",
    "def read_fasta(filename):\n",
    "    name, seq = None,''\n",
    "    fasta = open(filename, 'r')\n",
    "    for line in fasta:\n",
    "        if line.startswith('>') and name == None:\n",
    "            name = line.rstrip('\\n').replace('>','')\n",
    "        else:\n",
    "            if line.startswith('>') and name != None:\n",
    "                yield [name, seq]\n",
    "                name = line.rstrip('\\n').replace('>','')\n",
    "                seq = ''\n",
    "            else:\n",
    "                seq = seq + line.rstrip('\\n')\n",
    "    yield [name, seq]\n",
    "    fasta.close()\n",
    "\n",
    "def format_name(name):\n",
    "    name = name.split(';')[0]\n",
    "    name = re.sub(r\"_contig\\d*$|\",\"\", name)\n",
    "    name = re.sub(r\"_circular|\",\"\", name)\n",
    "    return name\n",
    "\n",
    "# create empty dictionary to hold gene (keys) and samples (values)\n",
    "#dict_genes = {}\n",
    "list_genes = []\n",
    "\n",
    "\n",
    "# list files in input dir\n",
    "for file in os.listdir(extract_alignments_dir):\n",
    "    # files ending with \".fasta\"\n",
    "    if file.endswith(\"fasta\"):\n",
    "        # get gene from file name\n",
    "        gene = file.replace(\".fasta\",\"\")\n",
    "        # read fasta\n",
    "        fasta = read_fasta(os.path.join(extract_alignments_dir, file))\n",
    "        #iterate through sequences in fasta\n",
    "        for name, seq in fasta:\n",
    "            contig, start_stop, strand, name = name.split(\";\")\n",
    "            start, stop = start_stop.split(\"-\")\n",
    "            # get sample name from contig\n",
    "            sample = format_name(contig)\n",
    "            list_genes.append([sample, contig, start, stop, strand, gene])\n",
    "            \n",
    "            ## if long assembly name found, reformat to sample only\n",
    "            #if re.search(\"contig\", name) or re.search(\"circular\", name):\n",
    "            #    name = format_name(name)\n",
    "            ## add gene and name to dictionary\n",
    "            #if dict_genes.get(gene) == None:\n",
    "            #    dict_genes[gene] = [name]\n",
    "            #else:\n",
    "            #    dict_genes[gene].append(name)\n",
    "\n",
    "dat_genes = pd.DataFrame(list_genes, columns=[\"ID\", \"Contig\", \"Start\", \"End\", \"Strand\", \"Name\"])\n",
    "dat_genes[\"Present_mafft_filtered\"] = \"True\"\n",
    "#dat_genes[\"Start\"] = dat_genes[\"Start\"].astype(float)\n",
    "#dat_genes[\"End\"] = dat_genes[\"End\"].astype(float)\n",
    "#dat_genes.drop(\"Start\", axis=1, inplace=True)\n",
    "#dat_genes.drop(\"End\", axis=1, inplace=True)\n",
    "dat_genes\n",
    "#dict_genes\n",
    "\n",
    "# get dict_genes sorted by key\n",
    "#dict_genes_keys = list(dict_genes.keys())\n",
    "#dict_genes_keys.sort()\n",
    "#dict_genes_sorted = {i: dict_genes[i] for i in dict_genes_keys}\n",
    "\n",
    "#dict_genes_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "67851a39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>4</td>\n",
       "      <td>1502</td>\n",
       "      <td>nad5_0</td>\n",
       "      <td>237982419.4</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>1586</td>\n",
       "      <td>2922</td>\n",
       "      <td>nad4</td>\n",
       "      <td>214655501.8</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>2921</td>\n",
       "      <td>3206</td>\n",
       "      <td>nad4l</td>\n",
       "      <td>4178913.3</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>3369</td>\n",
       "      <td>3867</td>\n",
       "      <td>nad6</td>\n",
       "      <td>3279670.2</td>\n",
       "      <td>+</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>3895</td>\n",
       "      <td>5047</td>\n",
       "      <td>cob</td>\n",
       "      <td>323602426.3</td>\n",
       "      <td>+</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>8669</td>\n",
       "      <td>9347</td>\n",
       "      <td>atp6</td>\n",
       "      <td>35353613.3</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>922</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>9340</td>\n",
       "      <td>9502</td>\n",
       "      <td>atp8</td>\n",
       "      <td>65220.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>925</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>9603</td>\n",
       "      <td>10317</td>\n",
       "      <td>cox2</td>\n",
       "      <td>107385243.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>10379</td>\n",
       "      <td>11915</td>\n",
       "      <td>cox1</td>\n",
       "      <td>566136198.5</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>931</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>12111</td>\n",
       "      <td>13056</td>\n",
       "      <td>nad2</td>\n",
       "      <td>61252546.4</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>397 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Contig  Start    End    Name        Score  \\\n",
       "0    Anartia_jatrophae_saturata_contig0      4   1502  nad5_0  237982419.4   \n",
       "2    Anartia_jatrophae_saturata_contig0   1586   2922    nad4  214655501.8   \n",
       "3    Anartia_jatrophae_saturata_contig0   2921   3206   nad4l    4178913.3   \n",
       "6    Anartia_jatrophae_saturata_contig0   3369   3867    nad6    3279670.2   \n",
       "7    Anartia_jatrophae_saturata_contig0   3895   5047     cob  323602426.3   \n",
       "..                                  ...    ...    ...     ...          ...   \n",
       "921       Catacroptera_cloanthe_contig0   8669   9347    atp6   35353613.3   \n",
       "922       Catacroptera_cloanthe_contig0   9340   9502    atp8      65220.0   \n",
       "925       Catacroptera_cloanthe_contig0   9603  10317    cox2  107385243.0   \n",
       "927       Catacroptera_cloanthe_contig0  10379  11915    cox1  566136198.5   \n",
       "931       Catacroptera_cloanthe_contig0  12111  13056    nad2   61252546.4   \n",
       "\n",
       "    Strand                          ID  \n",
       "0        -  Anartia_jatrophae_saturata  \n",
       "2        -  Anartia_jatrophae_saturata  \n",
       "3        -  Anartia_jatrophae_saturata  \n",
       "6        +  Anartia_jatrophae_saturata  \n",
       "7        +  Anartia_jatrophae_saturata  \n",
       "..     ...                         ...  \n",
       "921      -       Catacroptera_cloanthe  \n",
       "922      -       Catacroptera_cloanthe  \n",
       "925      -       Catacroptera_cloanthe  \n",
       "927      -       Catacroptera_cloanthe  \n",
       "931      -       Catacroptera_cloanthe  \n",
       "\n",
       "[397 rows x 7 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "ad790e17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>Present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>4</td>\n",
       "      <td>1502</td>\n",
       "      <td>nad5_0</td>\n",
       "      <td>237982419.4</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>1586</td>\n",
       "      <td>2922</td>\n",
       "      <td>nad4</td>\n",
       "      <td>214655501.8</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>2921</td>\n",
       "      <td>3206</td>\n",
       "      <td>nad4l</td>\n",
       "      <td>4178913.3</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>3369</td>\n",
       "      <td>3867</td>\n",
       "      <td>nad6</td>\n",
       "      <td>3279670.2</td>\n",
       "      <td>+</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>3895</td>\n",
       "      <td>5047</td>\n",
       "      <td>cob</td>\n",
       "      <td>323602426.3</td>\n",
       "      <td>+</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>8669</td>\n",
       "      <td>9347</td>\n",
       "      <td>atp6</td>\n",
       "      <td>35353613.3</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>922</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>9340</td>\n",
       "      <td>9502</td>\n",
       "      <td>atp8</td>\n",
       "      <td>65220.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>925</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>9603</td>\n",
       "      <td>10317</td>\n",
       "      <td>cox2</td>\n",
       "      <td>107385243.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>10379</td>\n",
       "      <td>11915</td>\n",
       "      <td>cox1</td>\n",
       "      <td>566136198.5</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>931</th>\n",
       "      <td>Catacroptera_cloanthe_contig0</td>\n",
       "      <td>12111</td>\n",
       "      <td>13056</td>\n",
       "      <td>nad2</td>\n",
       "      <td>61252546.4</td>\n",
       "      <td>-</td>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>397 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Contig  Start    End    Name        Score  \\\n",
       "0    Anartia_jatrophae_saturata_contig0      4   1502  nad5_0  237982419.4   \n",
       "2    Anartia_jatrophae_saturata_contig0   1586   2922    nad4  214655501.8   \n",
       "3    Anartia_jatrophae_saturata_contig0   2921   3206   nad4l    4178913.3   \n",
       "6    Anartia_jatrophae_saturata_contig0   3369   3867    nad6    3279670.2   \n",
       "7    Anartia_jatrophae_saturata_contig0   3895   5047     cob  323602426.3   \n",
       "..                                  ...    ...    ...     ...          ...   \n",
       "921       Catacroptera_cloanthe_contig0   8669   9347    atp6   35353613.3   \n",
       "922       Catacroptera_cloanthe_contig0   9340   9502    atp8      65220.0   \n",
       "925       Catacroptera_cloanthe_contig0   9603  10317    cox2  107385243.0   \n",
       "927       Catacroptera_cloanthe_contig0  10379  11915    cox1  566136198.5   \n",
       "931       Catacroptera_cloanthe_contig0  12111  13056    nad2   61252546.4   \n",
       "\n",
       "    Strand                          ID  Present  \n",
       "0        -  Anartia_jatrophae_saturata    False  \n",
       "2        -  Anartia_jatrophae_saturata     True  \n",
       "3        -  Anartia_jatrophae_saturata     True  \n",
       "6        +  Anartia_jatrophae_saturata     True  \n",
       "7        +  Anartia_jatrophae_saturata     True  \n",
       "..     ...                         ...      ...  \n",
       "921      -       Catacroptera_cloanthe     True  \n",
       "922      -       Catacroptera_cloanthe     True  \n",
       "925      -       Catacroptera_cloanthe     True  \n",
       "927      -       Catacroptera_cloanthe     True  \n",
       "931      -       Catacroptera_cloanthe     True  \n",
       "\n",
       "[397 rows x 8 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def gene_present_in_dat_genes(dat_genes, id, contig, name):\n",
    "    x = dat_genes [\n",
    "        (dat_genes[\"ID\"] == id) &\n",
    "        (dat_genes[\"Contig\"] == contig) &\n",
    "        (dat_genes[\"Name\"] == name)\n",
    "        ]\n",
    "    if x.shape[0] > 0:\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "dat_bed[\"Present\"] = dat_bed.apply(lambda row: gene_present_in_dat_genes(dat_genes, row[\"ID\"], row[\"Contig\"], row[\"Name\"]), axis=1)\n",
    "\n",
    "dat_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "c2ee63eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>Present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>4</td>\n",
       "      <td>1502</td>\n",
       "      <td>nad5_0</td>\n",
       "      <td>2.379824e+08</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>6123</td>\n",
       "      <td>6860</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>7146</td>\n",
       "      <td>7248</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>3.800000e-03</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Anartia_jatrophae_saturata_contig0</td>\n",
       "      <td>15083</td>\n",
       "      <td>15275</td>\n",
       "      <td>nad5_1</td>\n",
       "      <td>9.939820e+04</td>\n",
       "      <td>-</td>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Protogoniomorpha_temora_contig0</td>\n",
       "      <td>1</td>\n",
       "      <td>607</td>\n",
       "      <td>cox3_0</td>\n",
       "      <td>8.055191e+07</td>\n",
       "      <td>-</td>\n",
       "      <td>Protogoniomorpha_temora</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>Protogoniomorpha_temora_contig0</td>\n",
       "      <td>15057</td>\n",
       "      <td>15333</td>\n",
       "      <td>cox3_1</td>\n",
       "      <td>2.041177e+07</td>\n",
       "      <td>-</td>\n",
       "      <td>Protogoniomorpha_temora</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>Parasarpa_zayla_circular</td>\n",
       "      <td>7257</td>\n",
       "      <td>7359</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>1.500000e-03</td>\n",
       "      <td>+</td>\n",
       "      <td>Parasarpa_zayla</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>Parasarpa_zayla_circular</td>\n",
       "      <td>7647</td>\n",
       "      <td>8392</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>Parasarpa_zayla</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>Junonia_villida_contig0</td>\n",
       "      <td>1951</td>\n",
       "      <td>2140</td>\n",
       "      <td>nad6</td>\n",
       "      <td>8.062231e+05</td>\n",
       "      <td>+</td>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>Junonia_villida_contig3</td>\n",
       "      <td>74</td>\n",
       "      <td>236</td>\n",
       "      <td>nad6</td>\n",
       "      <td>1.759792e+05</td>\n",
       "      <td>+</td>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>Moduza_procris_circular</td>\n",
       "      <td>7845</td>\n",
       "      <td>7947</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>6.200000e-03</td>\n",
       "      <td>+</td>\n",
       "      <td>Moduza_procris</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>Moduza_procris_circular</td>\n",
       "      <td>8233</td>\n",
       "      <td>8973</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>Moduza_procris</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>Mallika_jacksoni_circular</td>\n",
       "      <td>2516</td>\n",
       "      <td>2618</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>3.800000e-03</td>\n",
       "      <td>+</td>\n",
       "      <td>Mallika_jacksoni</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>Mallika_jacksoni_circular</td>\n",
       "      <td>2918</td>\n",
       "      <td>3647</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>Mallika_jacksoni</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>Hypolimnas_usambara_contig0</td>\n",
       "      <td>382</td>\n",
       "      <td>431</td>\n",
       "      <td>rrnS_1</td>\n",
       "      <td>7.200000e-08</td>\n",
       "      <td>+</td>\n",
       "      <td>Hypolimnas_usambara</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>Hypolimnas_usambara_contig0</td>\n",
       "      <td>551</td>\n",
       "      <td>576</td>\n",
       "      <td>rrnS_2</td>\n",
       "      <td>8.300000e-07</td>\n",
       "      <td>+</td>\n",
       "      <td>Hypolimnas_usambara</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>Hypolimnas_usambara_contig0</td>\n",
       "      <td>598</td>\n",
       "      <td>1002</td>\n",
       "      <td>rrnS_0</td>\n",
       "      <td>3.500000e-32</td>\n",
       "      <td>+</td>\n",
       "      <td>Hypolimnas_usambara</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>Hypolimnas_usambara_contig0</td>\n",
       "      <td>1297</td>\n",
       "      <td>1399</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>2.100000e-03</td>\n",
       "      <td>+</td>\n",
       "      <td>Hypolimnas_usambara</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>Hypolimnas_usambara_contig0</td>\n",
       "      <td>1680</td>\n",
       "      <td>2430</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>Hypolimnas_usambara</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>Auzakia_danava_contig0</td>\n",
       "      <td>65</td>\n",
       "      <td>90</td>\n",
       "      <td>rrnS_1</td>\n",
       "      <td>3.800000e-07</td>\n",
       "      <td>-</td>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>Auzakia_danava_contig0</td>\n",
       "      <td>209</td>\n",
       "      <td>258</td>\n",
       "      <td>rrnS_0</td>\n",
       "      <td>1.700000e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>Auzakia_danava_contig2</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>rrnS</td>\n",
       "      <td>1.500000e-11</td>\n",
       "      <td>+</td>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>Auzakia_danava_contig2</td>\n",
       "      <td>350</td>\n",
       "      <td>452</td>\n",
       "      <td>rrnL</td>\n",
       "      <td>1.700000e-03</td>\n",
       "      <td>+</td>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>Auzakia_danava_contig1</td>\n",
       "      <td>1</td>\n",
       "      <td>442</td>\n",
       "      <td>cob</td>\n",
       "      <td>5.168863e+07</td>\n",
       "      <td>-</td>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>Adelpha_iphiclus_circular</td>\n",
       "      <td>4261</td>\n",
       "      <td>4363</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>1.300000e-03</td>\n",
       "      <td>+</td>\n",
       "      <td>Adelpha_iphiclus</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>Adelpha_iphiclus_circular</td>\n",
       "      <td>4646</td>\n",
       "      <td>5380</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>Adelpha_iphiclus</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>Phaedyma_columella_circular</td>\n",
       "      <td>8284</td>\n",
       "      <td>8386</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>9.500000e-03</td>\n",
       "      <td>+</td>\n",
       "      <td>Phaedyma_columella</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>Phaedyma_columella_circular</td>\n",
       "      <td>8669</td>\n",
       "      <td>9427</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>Phaedyma_columella</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>Kallima_paralekta_circular</td>\n",
       "      <td>7622</td>\n",
       "      <td>7724</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>3.500000e-03</td>\n",
       "      <td>+</td>\n",
       "      <td>Kallima_paralekta</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658</th>\n",
       "      <td>Kallima_paralekta_circular</td>\n",
       "      <td>8035</td>\n",
       "      <td>8771</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>Kallima_paralekta</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745</th>\n",
       "      <td>Precis_pelarga_circular</td>\n",
       "      <td>55</td>\n",
       "      <td>80</td>\n",
       "      <td>rrnS_2</td>\n",
       "      <td>8.300000e-07</td>\n",
       "      <td>-</td>\n",
       "      <td>Precis_pelarga</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>746</th>\n",
       "      <td>Precis_pelarga_circular</td>\n",
       "      <td>201</td>\n",
       "      <td>250</td>\n",
       "      <td>rrnS_1</td>\n",
       "      <td>3.700000e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>Precis_pelarga</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>Precis_pelarga_circular</td>\n",
       "      <td>14846</td>\n",
       "      <td>15224</td>\n",
       "      <td>rrnS_0</td>\n",
       "      <td>1.700000e-36</td>\n",
       "      <td>-</td>\n",
       "      <td>Precis_pelarga</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>Salamis_cacta_contig0</td>\n",
       "      <td>2</td>\n",
       "      <td>1211</td>\n",
       "      <td>nad5_0</td>\n",
       "      <td>1.538068e+08</td>\n",
       "      <td>+</td>\n",
       "      <td>Salamis_cacta</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>Salamis_cacta_contig0</td>\n",
       "      <td>9038</td>\n",
       "      <td>9140</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>8.500000e-03</td>\n",
       "      <td>+</td>\n",
       "      <td>Salamis_cacta</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811</th>\n",
       "      <td>Salamis_cacta_contig0</td>\n",
       "      <td>9430</td>\n",
       "      <td>10168</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>+</td>\n",
       "      <td>Salamis_cacta</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821</th>\n",
       "      <td>Salamis_cacta_contig0</td>\n",
       "      <td>14840</td>\n",
       "      <td>15218</td>\n",
       "      <td>nad5_1</td>\n",
       "      <td>1.031879e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>Salamis_cacta</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>822</th>\n",
       "      <td>Doleschallia_melana_contig0</td>\n",
       "      <td>22</td>\n",
       "      <td>475</td>\n",
       "      <td>nad6_0</td>\n",
       "      <td>2.856118e+06</td>\n",
       "      <td>+</td>\n",
       "      <td>Doleschallia_melana</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>Doleschallia_melana_contig0</td>\n",
       "      <td>2716</td>\n",
       "      <td>3445</td>\n",
       "      <td>rrnL_0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>Doleschallia_melana</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828</th>\n",
       "      <td>Doleschallia_melana_contig0</td>\n",
       "      <td>3736</td>\n",
       "      <td>3838</td>\n",
       "      <td>rrnL_1</td>\n",
       "      <td>3.800000e-03</td>\n",
       "      <td>-</td>\n",
       "      <td>Doleschallia_melana</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>Doleschallia_melana_contig0</td>\n",
       "      <td>15255</td>\n",
       "      <td>15378</td>\n",
       "      <td>nad6_1</td>\n",
       "      <td>9.410250e+04</td>\n",
       "      <td>+</td>\n",
       "      <td>Doleschallia_melana</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>Litinga_cottini_contig0</td>\n",
       "      <td>2</td>\n",
       "      <td>341</td>\n",
       "      <td>nad6-b</td>\n",
       "      <td>5.210666e+05</td>\n",
       "      <td>+</td>\n",
       "      <td>Litinga_cottini</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>Litinga_cottini_contig0</td>\n",
       "      <td>2558</td>\n",
       "      <td>3289</td>\n",
       "      <td>rrnL-b</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-</td>\n",
       "      <td>Litinga_cottini</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>867</th>\n",
       "      <td>Litinga_cottini_contig0</td>\n",
       "      <td>3587</td>\n",
       "      <td>3689</td>\n",
       "      <td>rrnL-a</td>\n",
       "      <td>1.600000e-04</td>\n",
       "      <td>-</td>\n",
       "      <td>Litinga_cottini</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899</th>\n",
       "      <td>Litinga_cottini_contig0</td>\n",
       "      <td>15027</td>\n",
       "      <td>15189</td>\n",
       "      <td>nad6-a</td>\n",
       "      <td>1.520142e+05</td>\n",
       "      <td>+</td>\n",
       "      <td>Litinga_cottini</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Contig  Start    End    Name         Score  \\\n",
       "0    Anartia_jatrophae_saturata_contig0      4   1502  nad5_0  2.379824e+08   \n",
       "11   Anartia_jatrophae_saturata_contig0   6123   6860  rrnL_0  0.000000e+00   \n",
       "12   Anartia_jatrophae_saturata_contig0   7146   7248  rrnL_1  3.800000e-03   \n",
       "38   Anartia_jatrophae_saturata_contig0  15083  15275  nad5_1  9.939820e+04   \n",
       "76      Protogoniomorpha_temora_contig0      1    607  cox3_0  8.055191e+07   \n",
       "113     Protogoniomorpha_temora_contig0  15057  15333  cox3_1  2.041177e+07   \n",
       "132            Parasarpa_zayla_circular   7257   7359  rrnL_1  1.500000e-03   \n",
       "133            Parasarpa_zayla_circular   7647   8392  rrnL_0  0.000000e+00   \n",
       "156             Junonia_villida_contig0   1951   2140    nad6  8.062231e+05   \n",
       "157             Junonia_villida_contig3     74    236    nad6  1.759792e+05   \n",
       "212             Moduza_procris_circular   7845   7947  rrnL_1  6.200000e-03   \n",
       "213             Moduza_procris_circular   8233   8973  rrnL_0  0.000000e+00   \n",
       "338           Mallika_jacksoni_circular   2516   2618  rrnL_1  3.800000e-03   \n",
       "339           Mallika_jacksoni_circular   2918   3647  rrnL_0  0.000000e+00   \n",
       "407         Hypolimnas_usambara_contig0    382    431  rrnS_1  7.200000e-08   \n",
       "408         Hypolimnas_usambara_contig0    551    576  rrnS_2  8.300000e-07   \n",
       "409         Hypolimnas_usambara_contig0    598   1002  rrnS_0  3.500000e-32   \n",
       "411         Hypolimnas_usambara_contig0   1297   1399  rrnL_1  2.100000e-03   \n",
       "412         Hypolimnas_usambara_contig0   1680   2430  rrnL_0  0.000000e+00   \n",
       "484              Auzakia_danava_contig0     65     90  rrnS_1  3.800000e-07   \n",
       "485              Auzakia_danava_contig0    209    258  rrnS_0  1.700000e-08   \n",
       "504              Auzakia_danava_contig2      0     63    rrnS  1.500000e-11   \n",
       "506              Auzakia_danava_contig2    350    452    rrnL  1.700000e-03   \n",
       "512              Auzakia_danava_contig1      1    442     cob  5.168863e+07   \n",
       "534           Adelpha_iphiclus_circular   4261   4363  rrnL_1  1.300000e-03   \n",
       "535           Adelpha_iphiclus_circular   4646   5380  rrnL_0  0.000000e+00   \n",
       "622         Phaedyma_columella_circular   8284   8386  rrnL_1  9.500000e-03   \n",
       "623         Phaedyma_columella_circular   8669   9427  rrnL_0  0.000000e+00   \n",
       "657          Kallima_paralekta_circular   7622   7724  rrnL_1  3.500000e-03   \n",
       "658          Kallima_paralekta_circular   8035   8771  rrnL_0  0.000000e+00   \n",
       "745             Precis_pelarga_circular     55     80  rrnS_2  8.300000e-07   \n",
       "746             Precis_pelarga_circular    201    250  rrnS_1  3.700000e-08   \n",
       "783             Precis_pelarga_circular  14846  15224  rrnS_0  1.700000e-36   \n",
       "784               Salamis_cacta_contig0      2   1211  nad5_0  1.538068e+08   \n",
       "810               Salamis_cacta_contig0   9038   9140  rrnL_1  8.500000e-03   \n",
       "811               Salamis_cacta_contig0   9430  10168  rrnL_0  0.000000e+00   \n",
       "821               Salamis_cacta_contig0  14840  15218  nad5_1  1.031879e+07   \n",
       "822         Doleschallia_melana_contig0     22    475  nad6_0  2.856118e+06   \n",
       "827         Doleschallia_melana_contig0   2716   3445  rrnL_0  0.000000e+00   \n",
       "828         Doleschallia_melana_contig0   3736   3838  rrnL_1  3.800000e-03   \n",
       "860         Doleschallia_melana_contig0  15255  15378  nad6_1  9.410250e+04   \n",
       "861             Litinga_cottini_contig0      2    341  nad6-b  5.210666e+05   \n",
       "866             Litinga_cottini_contig0   2558   3289  rrnL-b  0.000000e+00   \n",
       "867             Litinga_cottini_contig0   3587   3689  rrnL-a  1.600000e-04   \n",
       "899             Litinga_cottini_contig0  15027  15189  nad6-a  1.520142e+05   \n",
       "\n",
       "    Strand                          ID  Present  \n",
       "0        -  Anartia_jatrophae_saturata    False  \n",
       "11       -  Anartia_jatrophae_saturata    False  \n",
       "12       -  Anartia_jatrophae_saturata    False  \n",
       "38       -  Anartia_jatrophae_saturata    False  \n",
       "76       -     Protogoniomorpha_temora    False  \n",
       "113      -     Protogoniomorpha_temora    False  \n",
       "132      +             Parasarpa_zayla    False  \n",
       "133      +             Parasarpa_zayla    False  \n",
       "156      +             Junonia_villida    False  \n",
       "157      +             Junonia_villida    False  \n",
       "212      +              Moduza_procris    False  \n",
       "213      +              Moduza_procris    False  \n",
       "338      +            Mallika_jacksoni    False  \n",
       "339      +            Mallika_jacksoni    False  \n",
       "407      +         Hypolimnas_usambara    False  \n",
       "408      +         Hypolimnas_usambara    False  \n",
       "409      +         Hypolimnas_usambara    False  \n",
       "411      +         Hypolimnas_usambara    False  \n",
       "412      +         Hypolimnas_usambara    False  \n",
       "484      -              Auzakia_danava    False  \n",
       "485      -              Auzakia_danava    False  \n",
       "504      +              Auzakia_danava    False  \n",
       "506      +              Auzakia_danava    False  \n",
       "512      -              Auzakia_danava    False  \n",
       "534      +            Adelpha_iphiclus    False  \n",
       "535      +            Adelpha_iphiclus    False  \n",
       "622      +          Phaedyma_columella    False  \n",
       "623      +          Phaedyma_columella    False  \n",
       "657      +           Kallima_paralekta    False  \n",
       "658      +           Kallima_paralekta    False  \n",
       "745      -              Precis_pelarga    False  \n",
       "746      -              Precis_pelarga    False  \n",
       "783      -              Precis_pelarga    False  \n",
       "784      +               Salamis_cacta    False  \n",
       "810      +               Salamis_cacta    False  \n",
       "811      +               Salamis_cacta    False  \n",
       "821      +               Salamis_cacta    False  \n",
       "822      +         Doleschallia_melana    False  \n",
       "827      -         Doleschallia_melana    False  \n",
       "828      -         Doleschallia_melana    False  \n",
       "860      +         Doleschallia_melana    False  \n",
       "861      +             Litinga_cottini    False  \n",
       "866      -             Litinga_cottini    False  \n",
       "867      -             Litinga_cottini    False  \n",
       "899      +             Litinga_cottini    False  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat_bed [ dat_bed[\"Present\"] == False ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ecc33433",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adelpha_iphiclus</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Anartia_jatrophae_saturata</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Araschnia_levana</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Auzakia_danava</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Baeotus_beotus</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Catacroptera_cloanthe</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Chalinga_pratti</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Diaethria_gabaza_eupepla</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Doleschallia_melana</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Eurema_blanda</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Hypolimnas_usambara</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Junonia_villida</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Kallima_paralekta</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Kallimoides_rumia</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Litinga_cottini</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Mallika_jacksoni</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Moduza_procris</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Parasarpa_zayla</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Phaedyma_columella</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Precis_pelarga</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Protogoniomorpha_temora</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Salamis_cacta</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Smyrna_blomfildia</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Tacola_larymna</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Yoma_algina</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            ID  Name\n",
       "0             Adelpha_iphiclus    15\n",
       "1   Anartia_jatrophae_saturata    15\n",
       "2             Araschnia_levana    15\n",
       "3               Auzakia_danava    13\n",
       "4               Baeotus_beotus    15\n",
       "5        Catacroptera_cloanthe    14\n",
       "6              Chalinga_pratti    15\n",
       "7     Diaethria_gabaza_eupepla    15\n",
       "8          Doleschallia_melana    15\n",
       "9                Eurema_blanda    15\n",
       "10         Hypolimnas_usambara    15\n",
       "11             Junonia_villida    13\n",
       "12           Kallima_paralekta    15\n",
       "13           Kallimoides_rumia    15\n",
       "14             Litinga_cottini    15\n",
       "15            Mallika_jacksoni    15\n",
       "16              Moduza_procris    15\n",
       "17             Parasarpa_zayla    15\n",
       "18          Phaedyma_columella    15\n",
       "19              Precis_pelarga    15\n",
       "20     Protogoniomorpha_temora    15\n",
       "21               Salamis_cacta    15\n",
       "22           Smyrna_blomfildia    15\n",
       "23              Tacola_larymna    15\n",
       "24                 Yoma_algina    15"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get gene counts per sample\n",
    "dat_genes.groupby([\"ID\"])[[\"Name\"]].count().reset_index()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python_general",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
